File: src-tauri/src/mcp/tools/planner.rs
use crate::error::AppError;
use crate::services::planner_service::{
    add_planner_draft_child, analyze_repository_for_planner, clear_planner_pending,
    confirm_planner_plan, create_planner_session, delete_planner_draft_node,
    rename_planner_draft_node, submit_planner_turn, submit_planner_voice_turn,
    update_planner_session,
};
use crate::state::AppState;
use serde_json::Value;

use super::action_args::ToolAction;
use super::action_result;

pub(super) async fn handle(state: &AppState, payload: Value) -> Result<Value, AppError> {
    let tool_action = ToolAction::parse(payload)?;
    let args = tool_action.args();

    match tool_action.action.as_str() {
        "create_planner_session" => action_result(
            "create_planner_session",
            create_planner_session(
                state.planner_service.clone(),
                &state.db,
                args.optional_string(&["provider_id", "providerId"])?,
                args.optional_string(&["model_name", "modelName"])?,
            )
            .await?,
        ),
        "update_planner_session" => action_result(
            "update_planner_session",
            update_planner_session(
                state.planner_service.clone(),
                &state.db,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.optional_string(&["provider_id", "providerId"])?,
                args.optional_string(&["model_name", "modelName"])?,
            )
            .await?,
        ),
        "clear_planner_pending" => action_result(
            "clear_planner_pending",
            clear_planner_pending(
                state.planner_service.clone(),
                &state.db,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
            )
            .await?,
        ),
        "submit_planner_turn" => action_result(
            "submit_planner_turn",
            submit_planner_turn(
                state.planner_service.clone(),
                state,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.required_string(&["user_input", "userInput"], "user_input")?,
                args.optional_string(&["selected_draft_node_id", "selectedDraftNodeId"])?,
                args.optional_string(&["product_id", "productId"])?,
            )
            .await?,
        ),
        "submit_planner_voice_turn" => action_result(
            "submit_planner_voice_turn",
            submit_planner_voice_turn(
                state.planner_service.clone(),
                state,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.required_string(&["transcript", "user_input", "userInput"], "transcript")?,
                args.optional_string(&["selected_draft_node_id", "selectedDraftNodeId"])?,
                args.optional_string(&["product_id", "productId"])?,
            )
            .await?,
        ),
        "confirm_planner_plan" => action_result(
            "confirm_planner_plan",
            confirm_planner_plan(
                state.planner_service.clone(),
                state,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
            )
            .await?,
        ),
        "rename_planner_draft_node" => action_result(
            "rename_planner_draft_node",
            rename_planner_draft_node(
                state.planner_service.clone(),
                &state.db,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.required_string(&["node_id", "nodeId"], "node_id")?,
                args.required_string(&["name"], "name")?,
            )
            .await?,
        ),
        "add_planner_draft_child" => action_result(
            "add_planner_draft_child",
            add_planner_draft_child(
                state.planner_service.clone(),
                &state.db,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.required_string(&["parent_node_id", "parentNodeId"], "parent_node_id")?,
                args.required_string(&["child_type", "childType"], "child_type")?,
                args.required_string(&["name"], "name")?,
                args.optional_string(&["summary"])?,
            )
            .await?,
        ),
        "delete_planner_draft_node" => action_result(
            "delete_planner_draft_node",
            delete_planner_draft_node(
                state.planner_service.clone(),
                &state.db,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.required_string(&["node_id", "nodeId"], "node_id")?,
            )
            .await?,
        ),
        "analyze_repository_for_planner" => action_result(
            "analyze_repository_for_planner",
            analyze_repository_for_planner(
                state.planner_service.clone(),
                &state.db,
                &state.artifact_base_path,
                args.required_string(&["session_id", "sessionId"], "session_id")?,
                args.required_string(&["repository_id", "repositoryId"], "repository_id")?,
                args.optional_string(&["selected_draft_node_id", "selectedDraftNodeId"])?,
                args.optional_string(&["product_id", "productId"])?,
            )
            .await?,
        ),
        other => Err(AppError::Validation(format!(
            "unsupported aruvi_planner action: {other}"
        ))),
    }
}

File: src/lib/tauri/planner.ts
import { invoke } from "./core";
import type {
  PlannerContactResult,
  PlannerDraftChildType,
  PlannerSessionInfo,
  PlannerTurnResponse,
  SpeechToTextResponse,
} from "../types";

// Planner commands
export const createPlannerSession = (data?: { providerId?: string; modelName?: string }) =>
  invoke<PlannerSessionInfo>("create_planner_session_command", {
    providerId: data?.providerId ?? null,
    provider_id: data?.providerId ?? null,
    modelName: data?.modelName ?? null,
    model_name: data?.modelName ?? null,
  });

export const updatePlannerSession = (data: { sessionId: string; providerId?: string; modelName?: string }) =>
  invoke<PlannerSessionInfo>("update_planner_session_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    providerId: data.providerId ?? null,
    provider_id: data.providerId ?? null,
    modelName: data.modelName ?? null,
    model_name: data.modelName ?? null,
  });

export const clearPlannerPending = (sessionId: string) =>
  invoke<PlannerSessionInfo>("clear_planner_pending_command", {
    sessionId,
    session_id: sessionId,
  });

export const submitPlannerTurn = (data: { sessionId: string; userInput: string; selectedDraftNodeId?: string | null; productId?: string | null }) =>
  invoke<PlannerTurnResponse>("submit_planner_turn_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    userInput: data.userInput,
    user_input: data.userInput,
    selectedDraftNodeId: data.selectedDraftNodeId ?? null,
    selected_draft_node_id: data.selectedDraftNodeId ?? null,
    productId: data.productId ?? null,
    product_id: data.productId ?? null,
  });

export const submitPlannerVoiceTurn = (data: { sessionId: string; transcript: string; selectedDraftNodeId?: string | null; productId?: string | null }) =>
  invoke<PlannerTurnResponse>("submit_planner_voice_turn_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    transcript: data.transcript,
    userInput: data.transcript,
    user_input: data.transcript,
    selectedDraftNodeId: data.selectedDraftNodeId ?? null,
    selected_draft_node_id: data.selectedDraftNodeId ?? null,
    productId: data.productId ?? null,
    product_id: data.productId ?? null,
  });

export const confirmPlannerPlan = (sessionId: string) =>
  invoke<PlannerTurnResponse>("confirm_planner_plan_command", {
    sessionId,
    session_id: sessionId,
  });

export const renamePlannerDraftNode = (data: {
  sessionId: string;
  nodeId: string;
  name: string;
}) =>
  invoke<PlannerTurnResponse>("rename_planner_draft_node_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    nodeId: data.nodeId,
    node_id: data.nodeId,
    name: data.name,
  });

export const addPlannerDraftChild = (data: {
  sessionId: string;
  parentNodeId: string;
  childType: PlannerDraftChildType;
  name: string;
  summary?: string;
}) =>
  invoke<PlannerTurnResponse>("add_planner_draft_child_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    parentNodeId: data.parentNodeId,
    parent_node_id: data.parentNodeId,
    childType: data.childType,
    child_type: data.childType,
    name: data.name,
    summary: data.summary ?? null,
  });

export const deletePlannerDraftNode = (data: {
  sessionId: string;
  nodeId: string;
}) =>
  invoke<PlannerTurnResponse>("delete_planner_draft_node_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    nodeId: data.nodeId,
    node_id: data.nodeId,
  });

export const analyzeRepositoryForPlanner = (data: {
  sessionId: string;
  repositoryId: string;
  selectedDraftNodeId?: string | null;
  productId?: string | null;
}) =>
  invoke<PlannerTurnResponse>("analyze_repository_for_planner_command", {
    sessionId: data.sessionId,
    session_id: data.sessionId,
    repositoryId: data.repositoryId,
    repository_id: data.repositoryId,
    selectedDraftNodeId: data.selectedDraftNodeId ?? null,
    selected_draft_node_id: data.selectedDraftNodeId ?? null,
    productId: data.productId ?? null,
    product_id: data.productId ?? null,
  });

export const transcribeAudio = (data: {
  providerId?: string;
  modelName?: string;
  audioBytesBase64: string;
  mimeType: string;
  locale?: string;
}) =>
  invoke<SpeechToTextResponse>("transcribe_audio_command", {
    providerId: data.providerId ?? null,
    provider_id: data.providerId ?? null,
    modelName: data.modelName ?? null,
    model_name: data.modelName ?? null,
    audioBytesBase64: data.audioBytesBase64,
    audio_bytes_base64: data.audioBytesBase64,
    mimeType: data.mimeType,
    mime_type: data.mimeType,
    locale: data.locale ?? null,
  });

export const speakTextNatively = (data: {
  text: string;
  voice?: string;
  locale?: string;
}) =>
  invoke<void>("speak_text_natively_command", {
    text: data.text,
    voice: data.voice ?? null,
    locale: data.locale ?? null,
  });

export const sendTwilioWhatsappMessage = (data: { to: string; content: string }) =>
  invoke<void>("send_twilio_whatsapp_message", {
    to: data.to,
    content: data.content,
  });

export const startTwilioVoiceCall = (data: { to: string; initialPrompt?: string }) =>
  invoke<void>("start_twilio_voice_call", {
    to: data.to,
    initial_prompt: data.initialPrompt ?? null,
  });

export const routePlannerContact = (data: {
  to: string;
  content: string;
  preferredChannel?: "whatsapp" | "voice";
  allowAfterHours?: boolean;
}) =>
  invoke<PlannerContactResult>("route_planner_contact_command", {
    to: data.to,
    content: data.content,
    preferred_channel: data.preferredChannel ?? null,
    allow_after_hours: data.allowAfterHours ?? null,
  });
